Allow alternate restarts after idle (#2187)
quicwg/base-drafts: draft-ietf-quic-recovery.md

pacer, implementers are referred to the Fair Queue packet scheduler (fq qdisc)
in Linux (3.11 onwards).

## Sending data after an idle period

> A sender is considered idle if it has no bytes in flight.  A sender can become

Reviewer: Or

```suggestion
 A sender becomes idle if it has no bytes in flight.  A sender can become
```

> when it is application limited and has no data to send, or when it marks packets

Reviewer 2: ```suggestion
 when it is application limited and has no data to send, or when it marks all packets
```

> as lost (and therefore not in flight).

Reviewer 2: ```suggestion
 as lost, and therefore not in flight.
```

Reviewer 2: Can this also happen when all packets are acknowledged?  I'm not sure what the preferred behavior is in that case.  But the current text now properly captures the "my last PTO was acked and everything else was declared lost."

Author: Yes, that is a fine case ... this is what we've always had, the RTO text was incorrect. Basically the sender goes idle as soon as bytes in flight goes to 0. Maybe that's all we need to say. I've removed more text to make this clearer.

> packets as lost (and therefore not in flight).

Author: The text that was there was probably leftover from a while ago... at any rate there is no PTO case. There is nothing marked as lost on a PTO, and bytes in flight does not change on a PTO (modulo the two packets that are sent in response).  So a sender cannot become idle on a PTO.

pacer, implementers are referred to the Fair Queue packet scheduler (fq qdisc)
in Linux (3.11 onwards).

## Sending data after an idle period

A sender is considered idle if it has no bytes in flight and no pending
ack-eliciting data to send.  A sender can become idle when it is application

> limited or when it encounters a retransmission timeout.

Author: I've replaced retransmission with probe. 
Actually I think this entire second point is wrong now. A sender does not become idle on a PTO, it might become idle once loss is established, when an ACK is received.

> limited or when it encounters a retransmission timeout.

Author: Also, idle is when there's nothing in flight. I don't think we need the "pending data to send" bit.

A sender's congestion window MUST not increase while it is idle.

> When a sender starts sending data after an idle period, it's sending rate SHOULD

Reviewer: ```suggestion
When a sender starts sending data after an idle period, its sending rate SHOULD
```

> be decided as follows.

Reviewer: ```suggestion
be decided as follows:
```

  - If the sender uses pacing, it does not need to reduce its congestion
    window. It SHOULD however pace the congestion window and MAY allow an

>     initial burst no larger than the initial congestion window.

Reviewer 3: I don't understand this. "MAY allow initial burst no larger", does this it MAY also allow larger?

Reviewer 3: better

  - If the sender does not use pacing, it SHOULD reset its congestion window to
    the smaller of the current congestion window and the initial congestion

>     window, as recommended for TCP (see Section 4.1 of {{?RFC5681}}).

Reviewer: Maybe ignore my suggestions and use this instead:

> When sending data after becoming idle, a sender MUST reset its congestion window to the initial congestion window (see blah), unless it paces the sending of packets.  A sender MAY retain its congestion window if it paces the sending of any packets in excess of the initial congestion window.

I realize that this changes the emphasis slightly, but I think that it's clearer and more succinct.

I still think that this is an abrupt transition, but I guess the answer to any complaint there is simply "pacing".

Reviewer 3: There is a degenerate case where traffic arrives in blocks just under init cwnd at a high rate. The application can then burst at a high frequency.

Author: I like this better. I'm using this.

the size of bursts sent into the network, the behavior when restarting from
idle depends upon whether pacing is used.

> If the sender uses pacing, the congestion window MUST not increase and does not

Reviewer 2: Possibly the "MUST not increase CWND" part should move a paragraph up, since it's really not pacing specific?

need to be reduced while the connection is idle.  When restarting from idle, a
paced connection SHOULD limit the initial burst of packets to no more than the
initial congestion window and subsequent packets SHOULD be paced.

A sender that does not use pacing SHOULD reset its congestion window to the
smaller of the current congestion window and the initial congestion window, as
recommended for TCP (see Section 4.1 of {{?RFC5681}}).

A sender MAY implement alternate mechanisms to update its congestion window

> after idle periods, such as those proposed for TCP in {{?RFC7661}}.

Reviewer 2: To be clear, we're not really expecting anyone to implement this, right?

Reviewer 3: As formulated this does not permit bursts up to a decayed cwnd because init cwnd is required if not pacing.

Author: We do want to limit bursts to init cwnd. Beyond that, instead of describing alt mechanisms here, I think RFC 7661 offers a good and well thought out set of considerations.

Reviewer 3: Right, I suspected as much, just wanted to be sure.
